src/mame/drivers/swtpc8212.cpp

```diff
@@ -22,82 +60,359 @@ class swtpc8212_state : public driver_device
 		: driver_device(mconfig, type, tag)
 		, m_maincpu(*this, "maincpu")
-		, m_chargen(*this, "chargen")
+		, m_pia0(*this, "pia0")
+		, m_pia1(*this, "pia1")
+		, m_uart(*this, "uart")
+		, m_crtc(*this, "crtc")
+		, m_chargen1(*this, "chargen1")
+		, m_chargen2(*this, "chargen2")
+		, m_video_ram(*this, "videoram")
+		, m_baud_rate(*this, "baud_rate")
+		, m_no_parity(*this, "no_parity")
+		, m_parity_select(*this, "parity_select")
+		, m_jumper_a(*this, "jumper_a")
+		, m_jumper_b(*this, "jumper_b")
+		, m_mode_switch(*this, "mode_switch")
+		, m_bell_timer(nullptr)
+		, m_beeper(*this, "beeper")
+		, m_printer(*this, "printer")
 	{
 	}
 
 	void swtpc8212(machine_config &mconfig);
 
-private:
-	MC6845_UPDATE_ROW(update_row);
+	enum
+	{
+		BELL_TIMER_ID = 1
+	};
 
+private:
 	void mem_map(address_map &map);
+	virtual void machine_reset() override;
+	virtual void machine_start() override;
+	virtual void device_timer(emu_timer &timer, device_timer_id id, int param, void *ptr) override;
 
 	required_device<cpu_device> m_maincpu;
-	required_region_ptr<u8> m_chargen;
+	required_device<pia6821_device> m_pia0;
+	required_device<pia6821_device> m_pia1;
+	required_device<ins8250_device> m_uart;
+	required_device<mc6845_device> m_crtc;
+	required_region_ptr<u8> m_chargen1;
+	required_region_ptr<u8> m_chargen2;
+	required_shared_ptr<u8> m_video_ram;
+	required_ioport m_baud_rate;
+	required_ioport m_no_parity;
+	required_ioport m_parity_select;
+	required_ioport m_jumper_a;
+	required_ioport m_jumper_b;
+	required_ioport m_mode_switch;
+	emu_timer *m_bell_timer;
+	required_device<beep_device> m_beeper;
+	required_device<printer_image_device> m_printer;
+
+	MC6845_UPDATE_ROW(update_row);
+
+	DECLARE_WRITE8_MEMBER(latch_w);
+	uint8_t m_latch_data;
+
+	DECLARE_READ8_MEMBER(pia0_pa_r);
+	DECLARE_READ8_MEMBER(pia0_pb_r);
+	DECLARE_WRITE_LINE_MEMBER(pia0_ca2_w);
+
+	DECLARE_WRITE8_MEMBER(pia1_pa_w);
+	DECLARE_READ8_MEMBER(pia1_pb_r);
+	DECLARE_READ_LINE_MEMBER(pia1_ca1_r);
+	DECLARE_WRITE_LINE_MEMBER(pia1_ca2_w);
+
+	uint8_t m_kbd_data;
+	uint8_t m_printer_data;
+	uint8_t m_printer_data_ready;
+
+	void kbd_put(uint8_t data);
 };
 
+static INPUT_PORTS_START(swtpc8212)
+
+	PORT_START("baud_rate")
+	PORT_DIPNAME(0x1f, 0x19, "Baud Rate (Bits 0 to 4)")
+	PORT_DIPSETTING(0x04, "110")
+	PORT_DIPSETTING(0x0a, "300")
+	PORT_DIPSETTING(0x0d, "600")
+	PORT_DIPSETTING(0x0f, "1200")
+	PORT_DIPSETTING(0x12, "2400")
+	PORT_DIPSETTING(0x16, "4800")
+	PORT_DIPSETTING(0x18, "7200")
+	PORT_DIPSETTING(0x19, "9600")
+	PORT_DIPSETTING(0x1c, "19200")
+	PORT_DIPSETTING(0x1f, "38400")
+
+	PORT_START("no_parity")
+	PORT_DIPNAME(0x40, 0x00, "No Parity (Bit 6)")
+	PORT_DIPSETTING(0x00, "No Parity (On)")
+	PORT_DIPSETTING(0x40, "Parity (Off)")
+
+	PORT_START("parity_select")
+	PORT_DIPNAME(0x80, 0x00, "Parity Select (Bit 7)")
+	PORT_DIPSETTING(0x00, "Odd or Mark (On)")
+	PORT_DIPSETTING(0x80, "Even or Space (Off)")
```

Comment:
Please combine these into a single port, and use DIP locations to indicate their location rather than adding it to the name.

Reply:
It should be noted that with terminals whose parity options are restricted to even, odd or none, the manuals often list the "no parity" switch separately.